### ISSUE TITLE
Image size during inference

### ISSUE THREAD

**Issue by DirkKramer** [ultralytics/ultralytics#19552]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I am trying to understand how inference handles my image size.

I have a trained model:
```
    model.train(
        data=data, 
        epochs=100, 
        batch=-1,
        imgsz=640,
        amp=True,
        augment=False,
        device=0)
```

when I run inference on a stream of images from my camera with image size of 1920x1080 using:
```
    frame_bgr = cv2.resize(frame_bgr,(640,640))
    results  = self.detect_model.predict(frame_bgr)
```

i get the following inference times:
```
0: 640x640 8 Salmons, 20.9ms
Speed: 2.7ms preprocess, 20.9ms inference, 3.7ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 8 Salmons, 14.7ms
Speed: 2.6ms preprocess, 14.7ms inference, 2.3ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 9 Salmons, 14.6ms
Speed: 2.3ms preprocess, 14.6ms inference, 2.3ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 7 Salmons, 15.9ms
Speed: 2.8ms preprocess, 15.9ms inference, 2.3ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 8 Salmons, 15.0ms
Speed: 2.3ms preprocess, 15.0ms inference, 2.3ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 9 Salmons, 15.4ms
Speed: 2.4ms preprocess, 15.4ms inference, 2.6ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 7 Salmons, 21.4ms
Speed: 2.6ms preprocess, 21.4ms inference, 2.3ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 9 Salmons, 14.8ms
Speed: 2.4ms preprocess, 14.8ms inference, 2.8ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 7 Salmons, 20.4ms
Speed: 2.9ms preprocess, 20.4ms inference, 2.5ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 7 Salmons, 17.4ms
Speed: 2.8ms preprocess, 17.4ms inference, 2.7ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 7 Salmons, 16.8ms
Speed: 2.5ms preprocess, 16.8ms inference, 2.7ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 6 Salmons, 14.7ms
Speed: 2.3ms preprocess, 14.7ms inference, 2.3ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 6 Salmons, 14.1ms
Speed: 2.8ms preprocess, 14.1ms inference, 2.8ms postprocess per image at shape (1, 3, 640, 640)

```

when I remove the resize line and run on the original image I get the following times:

```
0: 384x640 7 Salmons, 50.0ms
Speed: 1.7ms preprocess, 50.0ms inference, 3.5ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 7 Salmons, 86.5ms
Speed: 2.9ms preprocess, 86.5ms inference, 3.4ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 7 Salmons, 15.7ms
Speed: 36.8ms preprocess, 15.7ms inference, 2.6ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 8 Salmons, 17.0ms
Speed: 1.7ms preprocess, 17.0ms inference, 2.8ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 8 Salmons, 17.4ms
Speed: 2.1ms preprocess, 17.4ms inference, 37.3ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 10 Salmons, 27.7ms
Speed: 37.8ms preprocess, 27.7ms inference, 3.7ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 8 Salmons, 90.6ms
Speed: 35.9ms preprocess, 90.6ms inference, 2.9ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 8 Salmons, 90.9ms
Speed: 1.9ms preprocess, 90.9ms inference, 4.3ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 7 Salmons, 15.3ms
Speed: 1.6ms preprocess, 15.3ms inference, 2.7ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 6 Salmons, 17.0ms
Speed: 1.9ms preprocess, 17.0ms inference, 2.6ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 7 Salmons, 55.5ms
Speed: 1.9ms preprocess, 55.5ms inference, 3.3ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 6 Salmons, 17.1ms
Speed: 2.3ms preprocess, 17.1ms inference, 3.2ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 6 Salmons, 19.3ms
Speed: 2.0ms preprocess, 19.3ms inference, 4.5ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 5 Salmons, 16.9ms
Speed: 2.0ms preprocess, 16.9ms inference, 3.3ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 5 Salmons, 16.0ms
Speed: 1.7ms preprocess, 16.0ms inference, 2.9ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 4 Salmons, 56.4ms
Speed: 36.3ms preprocess, 56.4ms inference, 5.8ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 4 Salmons, 88.1ms
Speed: 2.7ms preprocess, 88.1ms inference, 2.8ms postprocess per image at shape (1, 3, 384, 640)

0: 384x640 4 Salmons, 18.0ms
Speed: 1.8ms preprocess, 18.0ms inference, 4.5ms postprocess per image at shape (1, 3, 384, 640)
```

So when I resize manually I get significant better inference times.  Why is that? Is it because square image inference is faster then rectangular image inference?

Also shouldnt I just then resize the image beforehand and then scale the results back up to my original size manually?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @DirkKramer, thank you for your detailed question and for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can explore helpful resources on inference and other topics, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many common questions are addressed there.

### 🐛 Debugging

If this is a bug, please ensure you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further. Your provided inference comparison is very helpful already, but if possible, include any additional reproducible code snippets or key details that could help us replicate the issue as closely as possible.

### ❓ Question

For queries like this, providing as much information as possible, including any relevant differences between your model training setup, inference settings, and hardware specs, can greatly assist. You may also find the explanations in [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) helpful for optimizing performance.

### Community & Support

We encourage you to engage with the Ultralytics community too! For real-time conversation and technical discussions, check out:
- [Discord](https://discord.com/invite/ultralytics) 🎧  
- [Discourse](https://community.ultralytics.com/) for more in-depth, thread-based questions  
- [Subreddit](https://reddit.com/r/Ultralytics) for broader discussions and latest updates  

### Environment Verification

Ensure that you are using the latest `ultralytics` version with an up-to-date environment to rule out resolved issues. Update with the following:

```bash
pip install -U ultralytics
```

YOLO performs best in environments with compatible dependencies. Recommended environments include:

- **Free GPU notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  
- **Google Cloud**: [GCP Quickstart](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)  
- **AWS**: [AWS Quickstart](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)  
- **Docker**: [Docker Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>  

### CI Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  
Ultralytics CI runs daily to verify that YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) work seamlessly across a variety of platforms.

---

Lastly, keep in mind this is an automated response 🛠️, and an Ultralytics engineer will review your question soon to provide further assistance. Stay tuned 😊!

**Comment by Y-T-G** [comment]
If you're resizing manually, then you also need to take into account the latency for that for fair comparison. Also your resizing is wrong. Ultralytics uses Letterbox resizing. You're simply performing naive resizing where aspect ratio is not preserved.

If you're using Ultralytics, you shouldn't be resizing manually.

**Comment by Y-T-G** [comment]
From your logs, it seems like your code gets stuck at some points. Probably due to load or throttling.

**Comment by DirkKramer** [comment]
Yes, that is probably the case.  I just tested on an Orin AGX, and i dont observe the problem there.

Thanks for pointing me in the right direction.

**Comment by glenn-jocher** [comment]
Great to hear you've resolved the issue! For optimal inference speeds with YOLO11, we recommend using square input resolutions (like 640x640) through the native `imgsz` parameter in `model.predict()`, as this aligns with the model's optimized architecture. The Ultralytics inference pipeline automatically handles aspect ratio preservation with letterbox resizing while maintaining detection accuracy. 

If you need further speed optimizations, you might explore:
- Half-precision inference (`half=True`)
- [OpenVINO export](https://docs.ultralytics.com/integrations/openvino/) for Intel hardware acceleration
- Batch processing where applicable

Let us know if you need any clarification on these optimization paths! 🚀
